core/src/main/scala/kafka/cluster/Partition.scala

```diff
@@ -359,10 +361,14 @@ class Partition(val topicPartition: TopicPartition,
   // a false positive under min isr check, it has to check the leaderReplicaIdOpt again. Though it can still be affected
   // by ABA problems when leader->follower->leader, but it should be good enough for a metric.
   def isUnderMinIsr: Boolean = {
-    leaderLogIfLocal.exists { partitionState.isr.size < _.config.minInSyncReplicas } && isLeader
+    leaderLogIfLocal.exists{partitionState.isr.size < effectiveMinIsr(_) } && isLeader
   }
 
-  def isAtMinIsr: Boolean = leaderLogIfLocal.exists { partitionState.isr.size == _.config.minInSyncReplicas }
+  private def effectiveMinIsr(leaderLog: UnifiedLog): Int = {
+      leaderLog.config.minInSyncReplicas.min(remoteReplicasMap.size + 1)
```

Comment:
Just curious, did we ever have a case where the min ISR was larger than the number of replicas?

Reply:
The removed UT tries to set the min ISR larger than the replication factor.
Basically, we don't have any restraints over the minISR value. If someone set or has set the minISR wrongly, it will be wield availability issue after the upgrade.

Reply:
Yea, it makes sense to constrain the min ISR like this, I was just wondering if we actually saw this problem in practice. Either way, I'm fine with fixing this here.